### ISSUE TITLE
Additional proxy level configuration options

### ISSUE THREAD

**Issue by ardikabs** [envoyproxy/gateway#1718]
*Description*:

Currently, my company is interested to spike the Envoy Gateway as our Ingress Controller replacement, therefore we do some load testing and apparently need to configure additional options to the Envoy Proxy.
So, I would like to know the chance of providing additional Envoy proxy options such as `--concurrency`, referring to the Istio implementation such as [ProxyConfig](https://istio.io/latest/docs/reference/config/networking/proxy-config/).

I would appreciate it if you could inform me whether this enhancement is already planned for the next development phase. If not, I am eager to contribute to the project and would like to get in touch for further collaboration.

**Comment by zirain** [comment]
thanks @ardikabs , this's something EG must supported.

**Comment by zirain** [comment]
it's ok, let get it step by step.

**Comment by arkodg** [comment]
@ardikabs curious why you need to tune `concurrency`, by default `concurrency` is equal to the number of `hardware threads` https://www.envoyproxy.io/docs/envoy/latest/operations/cli
are you trying to reduce this number for high core machines ?

**Comment by ardikabs** [comment]
@arkodg
We are under a circumstance where everything needs to be controlled, including the resources that have been taken by the workload. Thus, when we are setting it with the necessary limits such as 4 cores, Envoy won't run with that value according to the `--concurrency` default value which will preserve all hardware threads in the machine. Following https://www.envoyproxy.io/docs/envoy/latest/faq/performance/how_to_benchmark_envoy, we need to set the concurrency to match the number of cores/threads that are being reserved.

**Comment by arkodg** [comment]
@ardikabs have you tried to limit resources at a container level first, you can tweak the container resources for the EnvoyProxy https://gateway.envoyproxy.io/v0.4.0/api/config_types.html#kubernetescontainerspec ?

**Comment by arkodg** [comment]
@ardikabs if im understanding correctly, the container cpu requests/limits are set to 4, but the hardware threads are much higher so `concurrency` is needed to improve thread scheduling/performance of Envoy ?

**Comment by ardikabs** [comment]
@arkodg 
> have you tried to limit resources at a container level first, you can tweak the container resources for the EnvoyProxy https://gateway.envoyproxy.io/v0.4.0/api/config_types.html#kubernetescontainerspec ?

We have done this, but apparently, during the performance test, we got slightly different results, with `concurrency`, the result was slightly better in terms of utilization, given that the same RPM was reached but the utilization was slightly lower.
The setup was Envoy Proxy with Guaranteed QoS 4 core and 4 GiB, with concurrency set to the same size of resources (4 core) 70-72% CPU, but without concurrency being set the utilization getting 78-79%, thus we believe that predefined concurrency will help us get a better result.

> if im understanding correctly, the container cpu requests/limits are set to 4, but the hardware threads are much higher so concurrency is needed to drop Envoy worker thread count to improve thread scheduling/context-switch/performance of Envoy ?

Yes, I'm kinda new to this Envoy world, but apparently, the test result giving us this assumption, please CMIIW.

**Comment by arkodg** [comment]
@ardikabs thanks for the comment, dug a little deeper, in Envoy the number of worker threads can be possibly reduced by enabling `--cpuset-threads` thats based off the cpuset size which may be lesser than total hardware threads, but on my local k8s setup, I can't find a way to reduce cpusets for my Envoy pod, so end up with cpuset == hardware thread size == num envoy worker threads

TLDR, `--concurrency` is probably needed for fine tuning cpu performance
